tests/e2e/apps/hop/app.go

```
@@ -0,0 +1,253 @@
+	jsonStr, err := a.marshaller.MarshalToString(resp)
+	if err != nil {
+		glog.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write([]byte(jsonStr))
+	if err != nil {
+		glog.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	glog.Infof("Successfully served message %s at position %d", req.GetId(), p)
+}
+
+// Hop start a gRPC server for Hop App
+func (a App) Hop(ctx context.Context, req *config.HopMessage) (*config.HopMessage, error) {
+	p := req.GetPosition()
+	glog.Infof("GRPC Serving message %s at position %d", req.GetId(), p)
+	resp := a.forwardMessage(req)
+	glog.Infof("Successfully served message %s at position %d", req.GetId(), p)
+	return resp, nil
+}
+
+func (a App) makeRequest(m *config.HopMessage, d string) (*config.HopMessage, error) {
+	// Check destination and send using grpc or http handler
+	if strings.HasPrefix(d, "http://") {
+		return a.makeHTTPRequest(m, d)
+
+	} else if strings.HasPrefix(d, "grpc://") {
+		address := d[len("grpc://"):]
+		return a.makeGRPCRequest(m, address)
+	}
+	return nil, errors.New("protocol not supported")
+}
+
+func (a App) setNextPosition(m *config.HopMessage) {
+	p := m.GetPosition()
+	if p < 0 {
+		return
+	}
+	if m.GetRemoteDests()[p].GetError() != "" {
+		// Error along the way no need to continue
+		m.Position = -1
+		return
+	}
+	nextPos := m.GetPosition() + 1
+	if nextPos < int64(len(m.GetRemoteDests())) {
+		m.Position = nextPos
+		return
+	}
+	m.Position = -1
+}
+
+// MakeRequest will create a config.HopMessage proto and
+// send requests to defined remotes hosts in a chain.
+// Each Server is a client and will forward the call to the next hop.
+func (a App) MakeRequest(remotes *[]string) (*config.HopMessage, error) {
+	req := newHopMessage(remotes)
+	resp := a.forwardMessage(req)
+	var err error
+	for _, r := range resp.GetRemoteDests() {
+		if r.GetError() != "" {
```

Do you want to accumulate errors or return the first error found? The former requires something like https://github.com/hashicorp/go-multierror. For the latter you can return the error immediately.